### ISSUE TITLE
NetJavaImpl.HttpClientResponse is not implemented correctly

### ISSUE THREAD

**Issue by czyzby** [libgdx/libgdx#4700]
#### Issue details
`HttpClientResponse` silently ignores all exceptions thrown by the `Connection` input stream, including exception thrown if the connection was already closed. After the response content is extracted with `getResult` (or one of its alternatives), the input stream is closed and subsequent calls to `getResult` will return an empty byte array, empty stream or empty string, without warning the user.

Possible solutions:
- Cache response result as byte array. Wrap with string or input stream on demand. Make `HttpClientResponse` thread-safe and reusable.
- Rethrow silently ignored exceptions. I think making `HttpResponse` a single-use object is somewhat OK, as long as the user is able to track the source of the issue.
- At the very least - add some warning logs for ignored exceptions.

#### Reproduction
A Kotlin example reproducing the issue - should be trivial to convert to Java:
```Kotlin
fun main(args: Array<String>) {
  val net = NetJavaImpl()
  val request = HttpRequest("GET")
  request.url = "http://google.com"
  net.sendHttpRequest(request, object : HttpResponseListener {
    override fun cancelled() = Unit
    override fun failed(t: Throwable?) {
      t?.printStackTrace()
    }
    override fun handleHttpResponse(httpResponse: HttpResponse) {
      println("Content: ${httpResponse.resultAsString}")
      println("Content: ${httpResponse.resultAsString}")
    }
  })
  Thread.sleep(1000L)
}
```

`"Content: "` is printed by the second `println` execution. No exceptions are thrown.

#### Version of LibGDX
1.9.6, although I'm pretty sure the issue exists in earlier versions.

#### Please select the affected platforms
- [x] Android
- [x] iOS (robovm)
- [x] iOS (MOE)
- [ ] HTML/GWT
- [x] Windows
- [x] Linux
- [x] MacOS

**Comment by Tom-Ski** [comment]
The documentation states those methods should only be used once per response, and they don't throw exceptions when called multiple times, so as far as I can tell this is not an issue.  Exceptions thrown by 

If that is not the issue you are reporting, please expand and what it is you expect to happen given the current documentation.

I agree that it would be nicer to cache the result for subsequent calls, even though it is stated in the documentation, its a little too easy to miss.

**Comment by czyzby** [comment]
Except it does not work as expected. Try running this snippet:

```Kotlin
import com.badlogic.gdx.ApplicationAdapter
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Net.*
import com.badlogic.gdx.backends.lwjgl.LwjglApplication
import com.badlogic.gdx.net.NetJavaImpl

class App : ApplicationAdapter() {
  override fun create() {
    val net = NetJavaImpl()
    val request = HttpRequest("GET")
    request.url = "http://example.com"
    net.sendHttpRequest(request, object : HttpResponseListener {
      override fun cancelled() = Unit
      override fun failed(t: Throwable?) {
        t?.printStackTrace()
      }

      override fun handleHttpResponse(httpResponse: HttpResponse) {
        Gdx.app.postRunnable {
          println("Content: ${httpResponse.resultAsString}")
        }
      }
    })
  }
}

fun main(args: Array<String>) {
  LwjglApplication(App())
}
```
It will **not** print the actual result on the main rendering thread.

Even if documented, it's very easy to miss (as you've noticed), counterintuitive, dangerous and actually inconsistent with the documentation. Empty string/stream can be returned if the _first_ content read call is made on another thread.

This is a poorly designed API in the first place. Is there a use case for returning empty string or input stream from the HTTP response? Why would anyone want the framework to silently ignore exceptions and return invalid output in case of wrong API usage? This is what runtime exceptions are for - I'd very much rather have to deal with a `GdxRuntimeException` than wonder why my HTTP request yields no result.

Caching would also be nice. `HttpResponse` interface could also extend `Disposable`, so you can clear the cache.

**Comment by Tom-Ski** [comment]
The goal isn't to silently ignore exceptions but to to pass them along to the failed method. Exceptions from the connection being closed and fetching the input stream should be rethrown when appropriate. Such as trying to obtain the input stream after the connection has already been closed.

Afaik a no content response is a valid response if that is what you are asking.

I'm all for a refactor of these classes, there are a quite a few gotchas here, would be nice for a cleanup.

**Comment by czyzby** [comment]
> The goal isn't to silently ignore exceptions (...)

- https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L53
- https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L69
- https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L87
- https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L117

> (...) but to to pass them along to the failed method.

Exceptions thrown during the connecting phase? Sure. Exceptions thrown _after_ `handleHttpResponse` call or in case of invalid API usage? I don't think these should be passed to the `failed` method. Or silenced.

**Comment by Tom-Ski** [comment]
You misunderstand, exceptions should be (re)thrown when appropriate, not passed to failed.

A goal of this api is to pass exceptions that occur within the response handler along to failed.

**Comment by czyzby** [comment]
I understood, but this is not consistent with the current implementation - hence the links list.

> exceptions should be (re)thrown when appropriate, not *always* passed to failed.

Agreed.

**Comment by Tom-Ski** [comment]
Regarding thread safe use, are you suggesting to make it possible to read the content from another thread, or simply protect against this?

**Comment by czyzby** [comment]
Depends. If you decide to cache the result, I'd probably just go with reading the whole content as byte array, lazily creating and caching its string representation on first content read call and creating a new `ByteArrayInputStream` on each stream read call. Thread safely would not be an issue, as the response would be more or less immutable (unless you modify the raw bytes array, but I'm assuming no one would do that knowingly).

If there's no cache, see if it's possible to introduce thread safely without hacks and unacceptable performance penalty during regular use (content read only within the listener). If not, a meaningful exception would be nice - `input stream is closed` is a bit cryptic.

**Comment by czyzby** [comment]
As a quick fix, you could change [these](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L69) [two](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/net/NetJavaImpl.java#L87) into `throw new GdxRuntimeException("Unable to read response content: HTTP response stream is closed. Content can be read only once from within the HttpResponseListener#handleHttpResponse method.", e)`.

**Comment by noblemaster** [comment]
Yes, there appear several things wrong with the implementation in regards to error handling.

Also, I don't believe `getInputStream` will ever return `null`:
 -  [URLConnection.getInputStream](https://docs.oracle.com/javase/8/docs/api/java/net/URLConnection.html#getInputStream--)

This will probably throw an `UnknownServiceException`. Also, I think rather than returning `connection.getErrorStream()`, returning the actual exception probably makes more sense as proposed. All the `getResult*` should probably throw `IOException` in that sense. Also, the `void failed (Throwable t)` method should probably be deprecated or removed altogether for `HttpResponseListener`.

**Comment by MrStahlfelge** [comment]
Is `failed` not called? Anyway, you are all right that the code looks a bit messy, but on the other hand, there are no known problems with it. I do a lot of HttpRequests in my games, and never had any problems (only on iOS, [but that's another thing](https://github.com/MrStahlfelge/lightblocks/blob/1983286c3ef9e99e0a9071e9835a6684aaf96aa9/ios/src/com/badlogic/gdx/backends/iosrobovm/MyAppDelegate.java#L15)). So I think there's nothing to win here.

**Comment by noblemaster** [comment]
I am using it too and it's working fine. Yes, `failed` is called but it's a little bit redundant? Accessing the **response** can fail too, so we might as well just return all errors via the response directly? 

I definitely needs a cleanup either way. If it's an error, we should throw the error not return a text/error stream. But yes, as long as there are not errors, it's fine :)
